### PR TITLE
Cache PNPM store in CI

### LINKED ISSUE #23695: [CI] Node.exe not recognized as valid Win32 application (Java Distribution IT, Quarkus IT)
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the [latest nightly release](https://github.com/keycloak/keycloak/releases/tag/nightly)

### Area

ci

### Describe the bug

Lately there were errors on windows in Java Distribution IT and Quarkus IT jobs with execution of  `pnpm install --frozen-lockfile --ignore-scripts` command.

Error:
` Error:  Failed to execute goal com.github.eirslett:frontend-maven-plugin:1.12.1:pnpm (pnpm-install) on project keycloak-js-adapter-jar: Failed to run task: 'pnpm install --frozen-lockfile --ignore-scripts' failed. java.io.IOException: Cannot run program "D:\a\keycloak\keycloak\js\node\node.exe" (in directory "D:\a\keycloak\keycloak\js\libs\keycloak-js"): CreateProcess error=193, %1 is not a valid Win32 application -> [Help 1]`

### Version

main

### Expected behavior

Quarkus IT and Java Distribution IT should be passing.

### Actual behavior

Error with Node.js in JavaDistribution IT and Quartkus IT jobs

`Error:  Failed to execute goal com.github.eirslett:frontend-maven-plugin:1.12.1:pnpm (pnpm-install) on project keycloak-js-adapter-jar: Failed to run task: 'pnpm install --frozen-lockfile --ignore-scripts' failed. java.io.IOException: Cannot run program "D:\a\keycloak\keycloak\js\node\node.exe" (in directory "D:\a\keycloak\keycloak\js\libs\keycloak-js"): CreateProcess error=193, %1 is not a valid Win32 application -> [Help 1]`

### How to Reproduce?

Not clear as this is flaky behaviour

### Anything else?

Reference action runs:
https://github.com/keycloak/keycloak/actions/runs/6389624189/job/17341584983
https://github.com/keycloak/keycloak/actions/runs/6258874791/job/16994303411
https://github.com/keycloak/keycloak/actions/runs/6324894358/job/17175683119
https://github.com/keycloak/keycloak/actions/runs/6313350974/job/17141779462
https://github.com/keycloak/keycloak/actions/runs/6300070614/job/17102703740

### PR DESCRIPTION
**Pull request by jonkoops**
Caches the [PNPM](https://pnpm.io/) package store on CI. This should speed up the build process and reduce the amount of request outside of GitHub infrastructure, which should also reduce network limiting on our runners.

This PR also includes the following changes:
- All projects now use PNPM as a package manager instead of NPM, reducing complexity and improving consistency.
- All projects now use the same Node.js installation in from the root directory, reducing the amount of total downloads.
- All projects now automatically install Node.js and PNPM so they can be built individually (this is from the same cache, so this should have little to no performance impact).
- The [`--prefer-offline` flag](https://pnpm.io/cli/install#--prefer-offline) has been added to `pnpm install` to reduce the amount of network requests.
- The Node.js, PNPM and frontend-maven-plugin dependencies have been upgraded to their respective latest versions.

Closes #23695

**Edit:** This PR originally cached the installation for the Node.js and PNPM binaries, however these are already cached in the `.m2` directory and copied over when `frontend-maven-plugin` is activated. I've since changed the implementation to only cache packages held in the PNPM store to reduce installation time.

### CODE REVIEW

File: docs/documentation/release-details
Comment (jonkoops) [inline docs/documentation/release-details:1]: This file was referencing NPM, but it seems that it was no longer used anywhere and the versions are incredibly outdated. This seems to be covered by other logic in the GitHub Actions, so I've chosen to remove it.

File: prod-arguments.json
Comment (jonkoops) [inline prod-arguments.json:1]: This file references some NPM related settings, but I could not find any usage of it. Presumably this is used for productionization, but perhaps it is no longer needed. I've removed this file, but I can restore it if needed.

ping @iankko @ASzc @martin-kanis @drichtarik

Comment (ASzc) [reply prod-arguments.json:1]: This is fine, it's a legacy file that is not used on the main or 22.x branches

**Comment by ASzc** [comment]
I'm going to need to look at this before it is merged. Every time there's a NPM/PNPM change, it breaks the product build. Potentially this could be easier for that, as essentially we're forcing the system NPM to be used, rather than whatever frontend installs

File: prod-arguments.json
Comment (jonkoops) [reply prod-arguments.json:1]: Thanks, I will consider this resolved then 👍

File: .github/actions/frontend-plugin-cache/action.yml
Comment (ahus1) [inline .github/actions/frontend-plugin-cache/action.yml:12]: For the maven-cache action, we're using a weekly cache key with no restore-key.

This avoids too many caches, as `pom.xml`s change frequently, and restore-keys can make the caches grow with time. 

Would you consider changing this to weekly caches as well?

With the Maven cache, we eventually added a `create-cache-if-it-doesnt-exist` parameter, as we would want to cache only the artifacts created in the build, and the re-use them in other workflows, as the other workflows download fewer dependencies, and would then upload only a subset of the information we want to cache.

**Review by ahus1 — CHANGES_REQUESTED** [review]
Thank you for taking on this task. The changes look good, I only have to concerns about the cache key and which part of the build should be "allowed" to upload the cache. Thanks!

File: .github/actions/frontend-plugin-cache/action.yml
Comment (jonkoops) [reply .github/actions/frontend-plugin-cache/action.yml]: In this case it was only referencing the root POM, but I am really only interested in change to the Node.js and PNPM versions. I've changed this so the cache is only restored if `node.version` and `pnpm.version` match. This should also remove the need to keep weekly caches, as it's much more fine grained already.

> With the Maven cache, we eventually added a `create-cache-if-it-doesnt-exist` parameter, as we would want to cache only the artifacts created in the build, and the re-use them in other workflows, as the other workflows download fewer dependencies, and would then upload only a subset of the information we want to cache.

Mhhh, yeah I wonder if this needs/should be done for PNPM dependencies. We can always optimize this later if it causes trouble.

**Comment by jonkoops** [comment]
This PR originally cached the installation for the Node.js and PNPM binaries, however these are already cached in the `.m2` directory and copied over when `frontend-maven-plugin` is activated. I've since changed the implementation to only cache packages held in the PNPM store to reduce installation time.

**Review by ASzc — APPROVED** [review]
Still LGTM from maven+product perspective

**Review by ahus1 — DISMISSED** [review]
Thank you for the change. I reviewed the code and it looks good. I re-triggered some failed builds which apparently failed to download node, assuming this was a network glitch.

**Comment by jonkoops** [comment]
Before we go ahead and merge this, I am considering that #23695 will be closed. And although I am convinced the changes in this PR might reduce the number of instances we see of this issue, I am not entirely convinced that it will resolve it completely, as the [underlying bug ](https://github.com/keycloak/keycloak/issues/23695#issuecomment-1748355186)  might exist still.

Perhaps we should donate some effort to https://github.com/eirslett/frontend-maven-plugin/issues/882 to get this resolved permanently.

**Comment by ASzc** [comment]
Getting the plugin's bug fixed seems like it will take months, and the only way I see to fix #23695 without fixing the plugin would be to implement our own download code for node/npm (at least within the CI workflows), and have frontend just use it. Kinda like the product build supplies its own.

**Comment by jonkoops** [comment]
Yeah, I don't see that bug being fixed any time soon. We might consider forking the `frontend-maven-plugin` and adding our fixes to it while we try to upstream things.

**Comment by jonkoops** [comment]
The failures on CI are due to a bug in `frontend-maven-plugin` that needs to be fixed (https://github.com/eirslett/frontend-maven-plugin/issues/1115). I've submitted a PR (https://github.com/eirslett/frontend-maven-plugin/pull/1116) to fix this, but if it stays out too long I will see if we can apply a workaround for now.

**Comment by jonkoops** [comment]
The PNPM not working on Windows situation has been resolved now that a new version of `frontend-maven-plugin` has been released which includes my fix. The problems with Node.js installations becoming corrupt should now also be resolved in this new version (see https://github.com/eirslett/frontend-maven-plugin/pull/1118).

So I think we're in a good spot now to get this merged, can I ask you for another review @ASzc @ahus1?

File: .github/actions/frontend-plugin-cache/action.yml
Comment (jonkoops) [reply .github/actions/frontend-plugin-cache/action.yml]: I'll consider this one resolved.

**Review by ahus1 — APPROVED** [review]
LGTM, thank you for handling these two upstream issues.

**Comment by ahus1** [comment]
@jonkoops - LGTM, pending an UI team approval.

**Review by ssilvert — APPROVED** [review]

**Pull request merged**
